### PR TITLE
Split multi.jl, Move distributed processing code into its own module.

### PR DESCRIPTION
**Pull request by amitmurthy**
Title says it all.

No user facing changes. Should merge it in sooner than later. Will make it easier to backport any bugfixes once 0.6 is released.

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
Great – definitely preferable to get this in. For the future, it seems like it's probably better to make the silent reorg PR the base and the breaking/potentially controversial PR on top of that. That way the reorg can get merged without any fuss and the breaking change can get decided on its own.

File: base/parallel/Parallel.jl [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+module Parallel
+
+# imports for extension
+import Base: getindex, wait, put!, take!, fetch, isready, push!, length,
+             hash, ==, connect, kill, serialize, deserialize, close
+
+# imports for use
+import Base: Process, Semaphore, JLOptions, AnyDict, buffer_writes, wait_connected,
```
Comment (tkelman) [inline base/parallel/Parallel.jl]: if not extending, would be better to do `using` for these

File: base/parallel/Parallel.jl [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+             VERSION_STRING, sync_begin, sync_add, sync_end, async_run_thunk,
+            binding_module, notify_error
+
+# The following are required but cannot be imported at this
+# time as they are included after Parallel in sysimg.jl
```
Comment (tkelman) [inline base/parallel/Parallel.jl]: should Parallel be later then? what depends on it?

File: base/exports.jl [hunk 3]
```diff
@@ -1150,38 +1155,6 @@ export
     TCPSocket,
     UDPSocket,
 
-# multiprocessing
-    addprocs,
-    asyncmap,
-    asyncmap!,
```
Comment (amitmurthy) [inline base/exports.jl]: Good catch. Will correct.

File: base/parallel/Parallel.jl (hunk 1 above)
Comment (amitmurthy) [inline base/parallel/Parallel.jl]: OK.

**Comment by tkelman** [comment]
I have a trivial little diff-reducer helper script I use for reviewing PR's like this one that are mostly code movement, makes the diff about 90% smaller: https://gist.github.com/ef1713f9a58305f801dd9e325e3a998d

File: base/parallel/Parallel.jl (hunk 2 above)
Comment (amitmurthy) [inline base/parallel/Parallel.jl]: I remember trying to do so and ran into some cross dependencies the last time around. Will check again if I can prune this list somewhat, else I'll just let it be for now.

Comment (amitmurthy) [inline base/parallel/Parallel.jl]: The dependency was loading.jl which has now been moved too.

**Comment by amitmurthy** [comment]
Merging this in a while if there are no objections.

**Comment by JeffBezanson** [comment]
:+1: 

Just bikeshedding here, but maybe we should call this `Multiprocessing` or `Distributed`, since we now have two forms of parallelism?

**Comment by amitmurthy** [comment]
My vote is for `Distributed`

**Comment by tkelman** [comment]
Are any other module names adjectives?

**Comment by amitmurthy** [comment]
I don't think being an adjective is an issue. `Distributed` is short for Distributed computing/processing. An argument could be made that new comers will ask "distributed what?". Since the name will be mentioned/talked about in the context of distributed processing, I don't believe it will be confusing.

**Comment by StefanKarpinski** [comment]
Definitely prefer `Distributed`.

**Comment by tkelman** [comment]
Except that it also works on a single node. Distributed makes it sound like you need a cluster.

**Comment by Sacha0** [comment]
A vote for `Multiprocessing`, regularly using multiprocessing on a large single node. Nice to avoid conflation of disjoint concepts. Best!

**Comment by StefanKarpinski** [comment]
Doing "distributed" processing on a single node is always an available degenerate case...

**Comment by tkelman** [comment]
"Optionally distributed multiprocessing" would be a more accurate name here, but a little long.

If no other module names are adjectives, then yes there's a naming consistency problem with using an adjective here.

**Comment by ViralBShah** [comment]
+1 for Distributed, since that is what the primitives and the programming model are for.

**Comment by tkelman** [comment]
> Doing "distributed" processing on a single node is always an available degenerate case...

Sure, as the trivial case of not doing anything in parallel, if your unit of work is nodes. The unit of work in our multiprocessing code isn't distributed nodes, it's processes, so you can do nontrivial parallelism on a single node. "You can add either local or remote Distributed workers" sounds a lot more confusing than "you can add either local or remote Multiprocessing workers."

**Pull request merged**